src/sync.h

```diff
@@ -345,53 +357,70 @@ class CSemaphoreGrant
     bool fHaveGrant;
 
 public:
-    void Acquire()
+    void Acquire() noexcept
     {
-        if (fHaveGrant)
+        if (fHaveGrant) {
             return;
+        }
         sem->wait();
         fHaveGrant = true;
     }
 
-    void Release()
+    void Release() noexcept
     {
-        if (!fHaveGrant)
+        if (!fHaveGrant) {
             return;
+        }
         sem->post();
         fHaveGrant = false;
     }
 
-    bool TryAcquire()
+    bool TryAcquire() noexcept
     {
-        if (!fHaveGrant && sem->try_wait())
+        if (!fHaveGrant && sem->try_wait()) {
             fHaveGrant = true;
+        }
         return fHaveGrant;
     }
 
-    void MoveTo(CSemaphoreGrant& grant)
+    // Disallow copy.
+    CSemaphoreGrant(const CSemaphoreGrant&) = delete;
+    CSemaphoreGrant& operator=(const CSemaphoreGrant&) = delete;
+
+    // Allow move.
+    CSemaphoreGrant(CSemaphoreGrant&& other) noexcept
+    {
+        sem = other.sem;
+        fHaveGrant = other.fHaveGrant;
+        other.fHaveGrant = false;
+    }
```

Comment:
Is that necessary? I find it unexpected to leave the semaphore. Usually moved-from objects remain in some empty but valid state, e.g. as if default constructed. A semaphore grant that has a semaphore in it is not that. Did something break when you set `other.sem = nullptr`? I tried this patch:

```diff
@@ -390,20 +390,22 @@ public:
     // Allow move.
     CSemaphoreGrant(CSemaphoreGrant&& other) noexcept
     {   
         sem = other.sem;
         fHaveGrant = other.fHaveGrant;
         other.fHaveGrant = false;
+        other.sem = nullptr;
     }
     
     CSemaphoreGrant& operator=(CSemaphoreGrant&& other) noexcept
     {   
         Release();
         sem = other.sem;
         fHaveGrant = other.fHaveGrant;
         other.fHaveGrant = false;
+        other.sem = nullptr;
         return *this;
     }
```

and both unit tests and functional tests passed. I think the moved-from semaphore grant is not used after the move and is destroyed shortly afterwards.

Reply:
And `ThreadOpenAddedConnections` creates a `CSemaphoreGrant` before it loops over `vInfo`. Maybe that causes problems for a subsequent item? This function seems to relate to manually added connections.

Most functional tests use the `onetry` command when they call `addnode`, which calls `OpenNetworkConnection` directly. Except for `rpc_net.py` which uses `add`, but it's never trying more than 1 node at a time.

I'm able to illustrate the crash by modifying the above test to add a second peer connection: https://github.com/Sjors/bitcoin/commit/0079c5143a4efc50a016e970bcedbbf6eca79007 (and lowering the wait in `ThreadOpenAddedConnections` and having the test wait a bit)

Reply:
Hmm, so the code in `CConnman::ThreadOpenAddedConnections()` acquires the grant once and then opens connections to all entries in `vInfo`. Is this intended? Looks like a bug to me. It might as well exceed `MAX_ADDNODE_CONNECTIONS`. Unrelated to this PR, but should it acquire the grant inside the `for (const AddedNodeInfo& info : vInfo) {` loop?

Reply:
Wouldn't [`grant.TryAcquire()`](https://github.com/sipa/bitcoin/blob/ee80555c30eb18bec8fc30ba4679f55464878074/src/net.cpp#L2872) inside the for loop fail and `break` once we reach `MAX_ADDNODE_CONNECTIONS`?

Reply:
@vasild Agree. The code is fine at the moment, because passing `&grant` into `OpenNetworkConnections` at most only releases our grant, without invalidating it. Changing it to invalidate the grant makes our next addnode attempt fail due to reusing an invalid object.

Reply:
Right, I did not see the `TryAcquire()` call :facepalm:, no bug in master.

Now the question that is relevant to this PR: what about moving the grant creation inside the `for` loop:

```diff
--- i/src/net.cpp
+++ w/src/net.cpp
@@ -2861,18 +2861,18 @@ std::vector<AddedNodeInfo> CConnman::GetAddedNodeInfo() const
 void CConnman::ThreadOpenAddedConnections()
 {   
     AssertLockNotHeld(m_unused_i2p_sessions_mutex);
     AssertLockNotHeld(m_reconnections_mutex);
     while (true)
     {
-        CSemaphoreGrant grant(*semAddnode);
         std::vector<AddedNodeInfo> vInfo = GetAddedNodeInfo();
         bool tried = false;
         for (const AddedNodeInfo& info : vInfo) {
             if (!info.fConnected) {
-                if (!grant.TryAcquire()) {
+                CSemaphoreGrant grant(*semAddnode, /*fTry=*/true);
+                if (!grant) {
                     // If we've used up our semaphore and need a new one, let's not wait here since while we are waiti
                     // the addednodeinfo state might change.
                     break;
                 }
```

and having the move semantic (more naturally) leave the moved-from object in a default-constructed state (have grant=false, sem=null)?

Reply:
Nice! I don't think my test is usable other than for manual testing. We currently can't mock the timeouts involved. We also can't configure the maximum connection count so it would need 10 test nodes, though that's doable. I'll open an issue.

Reply:
Updated test that demonstrates the crash issue has been fixed: https://github.com/Sjors/bitcoin/tree/2023/09/semaphore-fixed

Reply:
Issue to add a test: #28635